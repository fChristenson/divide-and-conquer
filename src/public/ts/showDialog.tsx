import { FooDialog } from "./FooDialog";
import React from "react";
import { BarDialog } from "./BarDialog";
import { BazDialog } from "./BazDialog";

// The original function.
export const showDialog = (
  showFooDialog: boolean,
  showBarDialog: boolean,
  showBazDialog: boolean
) => {
  const day = new Date().getDay();

  if (showFooDialog && day === 1) return <FooDialog open />;
  if (showBarDialog) return <BarDialog open />;
  if (showBazDialog) return <BazDialog open />;
};

interface IShowDialogV2Props {
  showFooDialog: boolean;
  showBarDialog: boolean;
  showBazDialog: boolean;
}

// A decision making function like this will require a lot of parameters.
// By making the input an object we can make it very flexible.
export const showDialogV2 = (props: IShowDialogV2Props) => {
  const day = new Date().getDay();

  if (props.showBazDialog && day === 1) return <BazDialog open />;
  if (props.showFooDialog) return <FooDialog open />;
  if (props.showBarDialog) return <BarDialog open />;
};

interface IShowDialogV3Props {
  showFooDialog: boolean;
  showBarDialog: boolean;
  showBazDialog: boolean;
  day: number;
}

// By moving the dynamic value outside of the function we make the function easy to test
export const showDialogV3 = (props: IShowDialogV3Props) => {
  if (props.showBazDialog && props.day === 1) return <BazDialog open />;
  if (props.showFooDialog) return <FooDialog open />;
  if (props.showBarDialog) return <BarDialog open />;
};

interface IShowDialogV4Props {
  showFooDialog: boolean;
  showBarDialog: boolean;
  showBazDialog: boolean;
  day: number;
}

/**
 * We want to avoid showDialogV4 from growing in complexity.
 * If we isolate the conditional logic to a helper function we can test them independently.
 * We should also add a link to the expected order of presedence.
 *
 * The spec can be found at http://docs.local/this-is-the-order-of-dialogs
 */
export const showDialogV4 = (props: IShowDialogV4Props) => {
  if (showFooDialog(props)) return <FooDialog open />;
  if (props.showBarDialog) return <BarDialog open />;
  if (props.showBazDialog) return <BazDialog open />;
};

export const showFooDialog = (props: IShowDialogV4Props) => {
  return props.showFooDialog && props.day === 1;
};
