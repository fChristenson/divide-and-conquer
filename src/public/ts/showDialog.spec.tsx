import { showDialogV4 } from "./showDialog";

describe("showDialogV4", () => {
  it("should show the FooDialog", () => {
    const props = {
      showFooDialog: true,
      showBarDialog: true,
      showBazDialog: true,
      day: 1,
    };
    const node = showDialogV4(props);
    expect(node).toMatchSnapshot();
  });

  it("should show the BarDialog", () => {
    const props = {
      showFooDialog: false,
      showBarDialog: true,
      showBazDialog: true,
      day: 1,
    };
    const node = showDialogV4(props);
    expect(node).toMatchSnapshot();
  });

  it("should show the BazDialog", () => {
    const props = {
      showFooDialog: false,
      showBarDialog: false,
      showBazDialog: true,
      day: 1,
    };
    const node = showDialogV4(props);
    expect(node).toMatchSnapshot();
  });

  it("should show undefined", () => {
    const props = {
      showFooDialog: false,
      showBarDialog: false,
      showBazDialog: false,
      day: 1,
    };
    const node = showDialogV4(props);
    expect(node).toMatchSnapshot();
  });
});
