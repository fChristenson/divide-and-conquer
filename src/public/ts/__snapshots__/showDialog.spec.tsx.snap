// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`showDialogV4 should show the BarDialog 1`] = `
<Unknown
  open={true}
/>
`;

exports[`showDialogV4 should show the BazDialog 1`] = `
<Unknown
  open={true}
/>
`;

exports[`showDialogV4 should show the FooDialog 1`] = `
<Unknown
  open={true}
/>
`;

exports[`showDialogV4 should show undefined 1`] = `undefined`;
